src/test/java/org/broadinstitute/hellbender/engine/GATKPathSpecifierUnitTest.java

```diff
@@ -286,6 +287,173 @@ public void testStdOut() throws IOException {
+                {"gs://bucket/aFile.bam", ".bam"},
+                {"gs://hellbender/test/resources/aFile.adam", ".adam"},
+                {"gs://hellbender/test/resources/aFile.fasta", ".fasta"},
+                {"http://bucket/aFile.bam?query=param", ".bam"},
+
+                // getExtension() returns ".gz", but this case also satisfies hasExtension(".fasta.gz")
+                {"aFile.fasta.gz", ".gz"},
+                // basename is ".fasta"!
+                {".fasta.gz", ".gz"},
+        };
+    }
+
+    @Test(dataProvider = "getExtensionTestCases")
+    public void testGetExtension(final String spec, final String expectedExtension) {
+        final GATKPathSpecifier pathSpec = new GATKPathSpecifier(spec);
+        final String actualExtension = pathSpec.getExtension();
+
+        Assert.assertEquals(actualExtension, expectedExtension);
+        // verify that hasExtension(getExtension()) is always true
+        Assert.assertTrue(pathSpec.hasExtension(actualExtension));
+    }
+
+    @DataProvider(name="negativeGetExtensionTestCases")
+    public Object[][] negativeGetExtensionTestCases() {
+        return new Object[][]{
+                // no extensions
+                {""},
+                {"/"},
+                {"."},
+                {"localFile"},
+                {"localFile."},
+                {"/localFile."},
+                {"gs://hellbender/test/resources"},
+                {"gs://hellbender/test/resources?query=param"},
+                {"gs://hellbender/test/resources/"},
+                {"gs://hellbender/test/resources/?query=param"},
+        };
+    }
+
+    @Test(dataProvider = "negativeGetExtensionTestCases", expectedExceptions={IllegalArgumentException.class})
+    public void testNegativeGetExtension(final String spec) {
+        new GATKPathSpecifier(spec).getExtension();
+    }
+
+    @DataProvider(name = "hasExtensionTestCases")
+    public Object[][] hasExtensionTestCases() {
+        return new Object[][]{
+                // input, extension satisfies "hasExtension"
+                {"localFile.bam", ".bam", true },
+                {"localFile.BAM", ".BAM", true },
+                {"localFile.BAM", ".bam", true },
+                {"localFile.bam", ".BAM", true },
+                {"/localFile.bam", ".bam", true },
+                {"gs://bucket/aFile.bam", ".bam", true },
+                {"gs://hellbender/test/resources/aFile.adam", ".adam", true },
+                {"gs://hellbender/test/resources/aFile.fasta", ".fasta", true },
+                {"http://bucket/aFile.bam?query=param", ".bam", true },
+
+                {"aFile.fasta.gz", ".gz", true },
+                {"aFile.fasta.gz", ".fasta.gz", true },
+                // basename is ".fasta"!
+                {".fasta.gz", ".gz", true },
+                {".fasta.gz", ".fasta.gz", true },
+
+                // no extensions
+                {"/", ".ext", false },
+                {".", ".ext", false },
```

Comment:
Could you add a test for the case ".".hasExtension(".") ?

Reply:
Hm, it doesn't. "." gets turned into a URI with an absolute path that ends in "/./".